stint/io.nim

```diff
@@ -82,6 +82,112 @@ func truncate*(num: Stint or StUint, T: typedesc[SomeInteger]): T {.inline.}=
 func toInt*(num: Stint or StUint): int {.inline, deprecated:"Use num.truncate(int) instead".}=
   num.truncate(int)
 
+func bigToSmall(result: var (UintImpl | IntImpl), x: auto) {.inline.} =
+  when bitsof(x) == bitsof(result):
+    when type(result) is type(x):
+      result = x
+    else:
+      result = convert[type(result)](x)
+  else:
+    bigToSmall(result, x.lo)
+
+func smallToBig(result: var (UintImpl | IntImpl), x: auto) {.inline.} =
+  when bitsof(x) == bitsof(result):
+    when type(result) is type(x):
+      result = x
+    else:
+      result = convert[type(result)](x)
+  else:
+    smallToBig(result.lo, x)
+
+template selectConverter(N, bits: static[int], c1, c2, c3, c4, c5: untyped) =
+  when N < bits:
+    when N <= 64: c1
+    else: c2
+  elif N > bits:
+    when bits <= 64: c3
+    else: c4
+  else: c5
+
+func stuint*(x: StUint, bits: static[int]): StUint[bits] {.inline.} =
+  ## unsigned int to unsigned int conversion
+  ## smaller to bigger bits conversion will have the same value
+  ## bigger to smaller bits conversion, the result is truncated
+  const N = bitsof(x.data)
+  template c1 {.used.} = result = stuint(x.data, bits)
+  template c2 {.used.} = smallToBig(result.data, x.data)
+  template c3 {.used.} = result = stuint(x.truncate(type(result.data)), bits)
+  template c4 {.used.} = bigToSmall(result.data, x.data)
+  template c5 {.used.} = result = x
+  selectConverter(N, bits, c1, c2, c3, c4, c5)
```

Comment:
Why is this template considered better than inlining the branches here?
It doesn't reduce the overall code size and it's certainly obscures the intent significantly.

Reply:
They are both ugly :smile:.

Reply:
I'll revert it, since no benefits using templates.